# This is a Markdown document

to learn more, go to this [Markdown cheatsheet](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)
# This is the title
mean
## This is a subtitle

$ X = \sqrt{Y * 2 + Z^2} $

![](https://icatcare.org/app/uploads/2018/07/Thinking-of-getting-a-cat.png)

![This is a cat!](https://icatcare.org/app/uploads/2018/07/Thinking-of-getting-a-cat.png)

hjhjjh

Another point

In [29]:
3 + 2

5

In [3]:
3 + 2

5

## Exercises: Using Python as a calculator

Using the operators in the table below, ask Python to solve these difficult math equations:

|Assign to a Variable | Add  | Subract | Multiply | Divide | Power | Integer Divide | Remainder after Division | 
|  :---------------:  | :---:| :-----: | :------: | :----: | :---: | :------------: | :----------------------: |
|         =           |  +   |    -    |    *     |   /    |   **  |       //       |           %              |

1. What is two plus three?

In [5]:
2 + 3

5

2. What is two times three?

In [6]:
2 * 3

6

3. What is two to the third power?

In [7]:
2 ** 3

8

3.5 How many (whole) times does 7 go into 100?

In [8]:
100 // 7

14

4. What is the remainder after dividing 3 from 2?

In [11]:
2 % 3

2

In [12]:
3 % 2

1

In [13]:
5 % 2

1

In [14]:
5 % 10

5

In [15]:
10 % 5

0

In [10]:
3 + 2 + 1 + 3

9

#### Let's See: Does Python do operations left-to-right, or follow the mathematical order of operations (addition and subtraction only after multiplication and division)?

Read each of the following math expressions to yourself and calculate (on paper, or in your head) what the answer shoud be.  Then check the answer by calculating it in Python.  Were they the same answer?

3 + 5 + 2

3 * 5 + 2

3 + 5 * 2

In [16]:
3 + 5 * 2

13

(3 * 5) + 2

In [17]:
(3 * 5) + 2

17

3 * (5 + 2)

In [18]:
((3 * 2) + ((3 - 1) - 10))

-2

In [19]:
(((3) + 2) - 1 * 10)

-5

What order is Python running the code?

# Python as a Logic Tester
## Bools

You can also ask Python logical questions (questions that have a "True" or "False" answer). 

| "Is Equal to" | "Is Greater than" | "Is Greater than or Equal to"|  "Is not Equal to" |
|  :----------: |  :----------:     |  :----------: |  :----------: |
| ==            |   >               |   >=          |  !=  |

Python will return "True" if the statement is true, "False" if False

```python
>>> 3 > 2
True

>>> 3 < 2
False

>>> 3 == 2
False
```

Using logical operators, decide if the following assertions are True or False:

1. One hundred is bigger than ten.

In [30]:
100 > 10

True

In [31]:
4 * 10 < 30

False

2. Thirty squared is not equal to nine-hundred.

In [21]:
30 ** 2 != 900

False

3. These two equations are equal:  
  - 30 - 5 * 2
  - 100 / 5

In [22]:
30 - 5 * 2 == 100 / 5

True

In [23]:
30 - 5 * 2

20

In [24]:
100 / 5

20.0

4. The sum of the numbers from 1 to 5 are at least 15.

In [25]:
1 + 2 +3 +4 + 5 >= 15

True

(optional). The length of all the DNA in a single adult human, unravelled end to end, is greater than the distance from the earth to the the moon.

## Python as a Data Collector
Python also has operators for collecting related data together.  Most of this course will revolve around the pros and cons of different ways of collecting data, but let's take a look at them:

| "tuple" (fixed sequence) | "list" (changeable sequence) | "str" (sequence of text characters) | "bytes" (sequence of bytes) |  "set" {mathematical set)  | "dict" (key-value mapping) | 
| :---------:| :----:    | :--------:    | :--------:    | :--------:    | :--------:    |
|  (1, 2, 3) | [1, 2, 3] | "123" '123' | b"123" b'123' | {1, 2, 3} | {1: 2, 3: 4}  |

#### Exercises: Making Collections

Make four difference types of sequences, each containing only the numbers from 1 to 5.

In [32]:
(1, 2, 3, 4, 5)

(1, 2, 3, 4, 5)

In [34]:
"12345"

' 12345'

In [41]:
[1, 2, 3, 4, 5]

[1, 2, 3, 4, 5]

In [42]:
b"12345"

b'12345'

Make a sequence containing 3 names of people in this class.

In [38]:
( "Negar", "Andrea", "Anjali")

('Negar', 'Andrea', 'Anjali')

In [43]:
[(1, 2), (3, 4)]

[(1, 2), (3, 4)]

In [44]:
("Nick", 2.3, True, [1, 2])

('Nick', 2.3, True, [1, 2])

In [47]:
len((1, 2, 3, 3))

4

Make a list of four animals, ordering them by size.

('elephant', 'cow', 'ant')

Make a mapping of classmates and colors, relating each classmate to their favorie color.

In [51]:
{"Joe": "green", "Jenn": "blue"}
{"Joe: green, Jenn: blue"}
{"green": "Joe", "blue": "Jenn", "blue": "Jeff"}
{"green": "Joe", "blue": ["Jenn", "Jeff"]}

{'blue': ['Jenn', 'Jeff'], 'green': 'Joe'}

Make a mapping of colors to names, relating each color to the people that like it

Collect the set of all letters in your first name.

List examples of each type of collection in Python.

Sequences, in particular, have some interesting properties can be concatenated together with the '+' operator:

```python
>>> "Hello" + "World"
"HelloWorld"
```

Try concatenating three different types of sequences (containing whatever you wish)

In [52]:
[1, 2, 3] + [4 ,5, 6]

[1, 2, 3, 4, 5, 6]

In [53]:
"Hello" + "World"

'HelloWorld'

In [56]:
"Hello " * 3

'Hello Hello Hello '

In [55]:
[1, 2, 3] * 3

[1, 2, 3, 1, 2, 3, 1, 2, 3]

In [62]:
type(12), type(12.5), type(-20), type("23"), type([1, 2, 3])

(int, float, int, str, list)

## Python as a Data Extractor

Data can be indexed/queried/extracted from collectoins using the square brackets: [ ]

In sequences, putting a number inside the the brackets extracts the nth (counting from zero) value
```python
>>> (1, 2, 3)[1]
2

>>> (1, 2, 3)[0]
1

>>> (1, 2, 3)[-1]
3
```

You can "slice" a sequence (get all from one index to another index) using the colon [:]
```python
>>> (10, 20, 30, 40, 50, 60)[1:3]
(20, 30)

>>> (10, 20, 30, 40, 50, 60)[:3]
(10, 20, 30)

>>> (10, 20, 30, 40, 50, 60)[3:]
(40, 50, 60)
```

If you want to slice in steps of a certain size:

```python
>>>(10, 20, 30, 40, 50, 60)[1:6:2]
(20, 40, 60)

>>>(10, 20, 30, 40, 50, 60)[::2]
(10, 30, 50)

>>>(10, 20, 30, 40, 50, 60)[::-1]
(60, 50, 40, 30, 20, 10)
```


In dicts, putting the key inside the brackets returns the associated value.
```python
>>> {'a': 3, 'b': 4}['a']
3

>>> {'a': 3, 'b': 4}['b']
4
```


In [70]:
[Nick, Anna]

NameError: ignored

In [71]:
(1, 2, 3

SyntaxError: ignored

In [67]:
(1, 2, 3)[-4]

IndexError: ignored

## Indexing Exercises

Using the example dataset *scores*, select only the described elements from each list:

0. The first score

In [72]:
(0.2, 0.3, 0.9, 1.1, 2.2, 2.9, 0.0, 0.7, 1.3, 0.3, 0.5, 0.1, 0.0)[0]

0.2

1. The third score

In [73]:
(0.2, 0.3, 0.9, 1.1, 2.2, 2.9, 0.0, 0.7, 1.3, 0.3, 0.5, 0.1, 0.0)[2]

0.9

2. The last score

In [74]:
(0.2, 0.3, 0.9, 1.1, 2.2, 2.9, 0.0, 0.7, 1.3, 0.3, 0.5, 0.1, 0.0)[-1]

0.0

3. The 2nd through 5th score

In [76]:
(0.2, 0.3, 0.9, 1.1, 2.2, 2.9, 0.0, 0.7, 1.3, 0.3, 0.5, 0.1, 0.0)[1:5]

(0.3, 0.9, 1.1, 2.2)

4. Every second score (the first, third, fifth, etc)

In [79]:
(0.2, 0.3, 0.9, 1.1, 2.2, 2.9, 0.0, 0.7, 1.3, 0.3, 0.5, 0.1, 0.0)[1::2]

(0.3, 1.1, 2.9, 0.7, 0.3, 0.1)

In [85]:
[1, 2, 3, 4, 5, 6, 7, 8, 9][1::3][0:-1]

[2, 5]

5. Every score after the 4th score

In [86]:
(0.2, 0.3, 0.9, 1.1, 2.2, 2.9, 0.0, 0.7, 1.3, 0.3, 0.5, 0.1, 0.0)[4:]

(2.2, 2.9, 0.0, 0.7, 1.3, 0.3, 0.5, 0.1, 0.0)

6. Every second score from the 2nd to the 8th.

In [87]:
(0.2, 0.3, 0.9, 1.1, 2.2, 2.9, 0.0, 0.7, 1.3, 0.3, 0.5, 0.1, 0.0)[1:8:2]

(0.3, 1.1, 2.9, 0.7)

## Python as an Abstractable Data Aggregator

Python can use named **functions** that turn data into something else.  By doing this repeatedly, in sequence, we can build data processing pipelines!

Functions in Python have the following syntax:

```
output = function(input)
```

First, let's look at some **built-in** functions to get a sense of how they work:  **min()**, **max()**, **sum()**, and **len()**:

```python
>>> min([1, 2, 3, 2])
1

>>> max([1, 2, 3, 2])
3

>>> sum([1, 2, 3, 2])
8

>>> len([1, 2, 3, 2])
4
```

### Exercises: 
Read the following lines of Python and predict their output.  Then, run the code to see if you were correct.

```python
min([3, 6, 5, 2])
```

In [91]:
min([3, 6, 5, 2])

2

```python
max([3, 6, 5, 2][2:])
```

In [0]:
max([3, 6, 5, 2][2:])

```python
min([max((2, 4, 1)), len((2, 4, 1)), min([2, 4, 1])])
```

In [0]:
min([max((2, 4, 1)), len((2, 4, 1)), min([2, 4, 1])])

```python
(min([1, 2]), max([3, 4, 5]))
```

In [92]:
(min([1, 2]), max([3, 4, 5]))

(1, 5)

```python
max([1, 2, 3, 4, 5, 6][1:3])
```

In [93]:
max([1, 2, 3, 4, 5, 6][1:3])

3

```python
len({3, 4, 5, 5, 4})
```

In [94]:
len({3, 4, 5, 5, 4})

3

```python
len({1: 2, 3: 4, 5: 6, 7: 8})
```

In [95]:
len({1: 2, 3: 4, 5: 6, 7: 8})

4

## Python as an Ecosystem of Functions

There are millions of functions that can be used in Python.  To access them, you first have to **import** their **packages/modules**.  Then you can specify the functions inside the packages using the following syntax:

```
import package
output = package.function(input)
```

Sometimes, packages have **subpackages** that contain functions, too:

```
output = package.subpackage.function(input)
```

If you only need one subpackage in a function, you can make your lines shorter by importing the subpackage directly:

```
from package import subpackage
output = subpackage.function(input)
```

To get a list of functions in a package, try using the **dir()** function.  To learn what the function does, try the **help()** function.

```python
import math
dir(math)
help(math.sqrt)
```


In [102]:
import math
dir(math)
help(math.tan)
math.degrees()

Help on built-in function tan in module math:

tan(...)
    tan(x)
    
    Return the tangent of x (measured in radians).



Let's try out the built-in **math** and **statistics** packages to do some calculating!

What is the mean of all the integers from 1 to 6?

In [0]:
import statistics
import math

In [111]:
statistics.mean([1, 2, 3, 4, 5, 6])

3.5

In [110]:
statistics.mean([1, 2, 3, 4, 4])

2.8

What is the standard deviation of the integers from 1 to 6?

In [114]:
statistics.stdev([1, 2, 3, 4, 5, 6])

1.8708286933869707

What is the square root of 72?

In [115]:
math.sqrt(72)

8.48528137423857

What is pi?

In [118]:
math.pi

3.141592653589793

What how many radians are in 360 degrees?

In [120]:
math.radians(360)

6.283185307179586

Can you find a package that might have functions for random numbers?  Use it to generate a couple random numbers!

In [132]:
import random
random.seed(2)
random.random(), random.random()

(0.9560342718892494, 0.9478274870593494)

Look for names of some more built-in packages online!  Try importing them here.

## Python as State Machine

Python can save all of its **objects** to your computer's internal memory by **assigning** a a **name** to it.  This is like giving the object a tag that can be referenced later, and it has many subtleties that we'll be exploring throughout the course.  Here's the syntax:

```
name = data
```

For example:

```python
x = 3
nick = 100
j_people = ["John", "Jeff", "Judy", "Jenny"]
x2 = 4
```

While it looks like a normal math expression, this is, in reality, a **statement**; it changes the state of the program.  Let's explore it here:



#### Exercises

What should **x** equal?
```python
x = 3
x = 4
```

In [136]:
globals()['x'] = 15
x

15

```python
x = 3
x = x + 1
x = x + 1
x = x + 1
```

```python
x = 3
x -= 1
```

In [138]:
x = 3
x -= 1
x

2

```python
x = "Nick"
x = 0
```

```python
max = sum
x = max([1, 2, 3])
```

In [0]:
mean = 4
import math
math.pi
pi
statistics.pi

In [139]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


```python
max = sum
x = sum([1, 2, 3])
```

```python
x = [1, 2, 3]
x[1] = 4
```

In [141]:
x = [1, 2, 3]
x[1] = 4
x

[1, 4, 3]

```python
x = [1, 2, 3]
x[-1] = 4
```

```python
x = [1, 2, 3]
x[0] = 4
```

In [145]:
x = (1, 2, 3)
x[0] = 4
x

TypeError: ignored

## Review: Class Discussion

  1. Python has a lot of symbols in its syntax.  What are they, and what are they used for?
  2. What are functions?  How do you use them?  Where can you get them?
  3. What kinds of data collections are built-in to Python?  What concepts do they represent?
  4. What is a "logical expression"?  What is a "state change"?
  5. When are we getting coffee?

# Break

In [148]:
name = "Nick"
str.upper(name)
name.upper()

'NICK'

# Exercise: Getting Help By Drawing With Turtles

# Built-In Type Methods

As we discussed in the lecture, all Python objects are **Objects**, which means they have **Properties/Attributes** (data) and **Methods** associated with them.  

|  Object    | type(Object)  |  Example Method  |  Example Attribute  | 
| :----:     | :----------:  |  :------------:  |  :---------------:  |
| "Hello"    |     str       |     lower()      |                     |
|    42      |     int       |    bit_length()  |     numerator       |
|   3.14     |    float      |    is_integer()  |     imag            |
| ['a', 'b'] |  list         |     append()     |                     |


These are accessible via the **. (dot)**, and an object's methods can be browsed in the notebook by typing the **Tab** key after the dot.

For example, here we use a **string** **Class**'s *title()* method to capitalize the first letter of each word of the film title:

In [0]:
film = 'the lion king'
film

'the lion king'

In [0]:
type(film)

str

In [0]:
bigfilm = film.title()
bigfilm

'The Lion King'

## Exercises: Strings

Use the string methods to modify the following strings to the versions described below.

### Changing Strings

Use the **strip**, **swapcase**, **title**, **lower**, **upper**, and **zfill** methods, or the **\+** or **\*** operators, to perform the following modifications:

**"I am angry"** -> **"I AM ANGRY"**

In [150]:
x = "I am angry"
x.upper()

'I AM ANGRY'

**"the lion king"** -> **"The Lion King"**

**He said, "Are you okay?"** -> **hE SAID, "aRE YOU OKAY?"**

**three** -> **threethreethree**

### Searching Substrings

Use the **replace**, **count**, and **index** methods to search for a substring and answer the following questions using the example sentence below:

```python
sentence = "The dog ate my watermelon when we were digging at the beach!"
```

In [0]:
sentence = "The dog ate my watermelon when we were digging at the beach!"

How many "w" characters are in this sentence?

What character index does the word "watermelon" start on? (the first? The tenth?)

Replace the word "dog" in the sentence with the word "cat"

### Formatting Strings

The most common method used for strings is the **format()** methods, so it deserves special mention.  It is used to insert data into a string.  It is similar to the **replace()** method but only replaces curly braces **{}**, but it is *way* more powerful.  In fact, there is a whole web page dedicated to showing all the ways you can use it: https://pyformat.info/

```python
label_fmt = "Sample {}: Vial {}"
label_fmt.format(3, 4)
>> "Sample 3: Vial 4"
```

Using the **format()** method, make the following three sentences out of the original sentence below:

In [0]:
sentence = "I bought {} dogs."

1. "I bought 3 dogs."

2. "Wendy bought 5 dogs."

3. "George bought 700 dogs."

## Lists

Lists represent sequences of data.  They are made with either the square brackets **[]** or the **list()** constructor function.  They can be appended to, just like strings, but they can contain any kind of data--numbers, strings, even other lists.  Their methods mostly revolve around inserting and removing things:

| Method   |  What it Does                                  | 
| :-----:  |  :----------:                                  |
| append() | Appends an item to the end of the list.        | 
| extend() | Appends several items to the end of the list.  | 
| remove() | Removes an item from the list.                 |
| clear()  | Removes everything from the list.              | 
| copy()   | Returns a copy of the list.                    |
| count()  | Removes a specified item from the list         | 
| sort()   | Sorts the list.                                |
| reverse()| Reverses the order of items in the list.       |

**Example**

Here is a list of colors:

In [0]:
colors = ['Red', 'Blue', 'Yello']
colors

['Red', 'Blue', 'Yello']

Let's append the color 'Yellow' to the list:

In [0]:
colors.append('Yellow')
colors

['Red', 'Blue', 'Yello', 'Yellow']

*Note*: If you run the above code multiple times, 'Yellow' will be appended to the list multiple times!

### List Method Exercises

1. Make a list of 3 movie titles.

2. You have a list of subjects for your experiment, and you want to add another one to that list!  How could you do it?

In [0]:
subjects = ['NOP234', 'GHS673', 'JGL212']
new_subject = 'ASF193'

3. Now, a bunch of new subjects appeared!  How do you extend/append them to the main list?

In [0]:
subjects = ['NOP234', 'GHS673', 'JGL212']
new_subjects = ['ASF193', 'THW994', 'JJZ231']

4. Please sort those subjects alphabetically.  It looks better that way, doesn't it?

In [0]:
subjects = ['NOP234', 'GHS673', 'jGL212', 'ASF193', 'THW994', 'JJZ231']

5. Oh, no, 'JGL202' was a terrible subject, he intentionally ruined your study.  Well, there's no way you're keeping him.  How do you remove him from the list?

In [0]:
subjects = ['NOP234', 'GHS673', 'JGL212', 'ASF193', 'THW994', 'JJZ231']

## Further Learning Resources

  - Official Documentation on String Methods: https://docs.python.org/3/library/stdtypes.html#string-methods
  
  - Online Tutorial on Python Strings: https://realpython.com/python-strings/
  
  - Official Documentation on List Methods: https://docs.python.org/3/library/stdtypes.html#lists
  
  - Online Tutorial on Python Lists: https://realpython.com/python-lists-tuples/
  
  

## Extra Exercises

If you'd like to do some more, here they are!


1. You want to label your figure title with the subject code, and the subject code changes depending on which subject is being shown!  How could you stick a subject's name in your title string?

In [0]:
subject = 'NOP234'
title = "Mean Values of SUBJECT NAME's data over Time"

2. If you had formatted your title string in the following way, though, another string method would be more useful.  Which method would you use for this title:

In [0]:
subject1 = 'NOP234'
subject2 = 'GHS673'
title = "Performance Comparison between Subjects {} and {}"

3. The conference says it only takes abstracts that have a maximum word count of 100 words.  Did our abstract make the cut?  

*Hint: the len() function is useful here*

In [0]:
abstract = """We analyze the locomotor behavior of the rat during exploration, and show that digitally collected data (time series of positions) provide a sufftcient basis for establishing that the rat uses several distinct modes of motion (first, second, third, and sometimes fourth gear). The distinction between these modes is obtained by first segmenting the time series into sequences of data points occurring between arrests (as ascertained within the resolution of the data acquisition system). The statistical distribution of the maximal amount of motion occurring within each of these episodes is then analyzed and shown to be multi modal. This enables us to decompose motion into distinct modes."""


Oh, wait, now that I look closer, it actually says that it's a maximum of 100 **unique** words--duplicated words don't count.  How many unique words do we have in our abstract?  (What a strange conference...)


*Hint: Try this out with the **set()** class.*